File: droplet.cpp

#include <math.h>
#include "droplet.hpp"

int Droplet::num_droplets = 0;

Droplet::Droplet() {
    _multi = 0;
    _rcubed = _solute = _density = 0.;

    _mass = _volume = 0.;

    num_droplets++;
}

Droplet::Droplet(long multi, double rcubed, double solute,
                 double density) {
    _multi = multi;
    _rcubed = rcubed;
    _solute = solute;
    _density = density;

    _volume = (4.*M_PI/3.)*_rcubed;
    _mass = _volume*_density;

    num_droplets++;
}

Droplet::Droplet(const Droplet &d) {
    _multi = d._multi;
    _rcubed = d._rcubed;
    _solute = d._solute;
    _density = d._density;

    _volume = d._volume;
    _mass = d._mass;

    num_droplets++;
}

Droplet::~Droplet() {
    --num_droplets;
}

double Droplet::get_mass() const {
    return _mass;
}

long Droplet::get_multi() const {
    return _multi;
}

double Droplet::get_radius() const {
    return pow(_rcubed, 1./3.);
}

double Droplet::get_solute() const {
    return _solute;
}

double Droplet::get_volume() const {
    return _volume;
}

double Droplet::get_terminal_v() const {
    // BEARD, 1976
    double mass = this->get_mass();

    double r = this->get_radius();
    double d = 2.*r*1e6; // diameter, m -> micron
    double x = mass * 1e3; // convert kg -> g

    double alpha, x_to_beta;

    if (d <= 134.43)
    {
        alpha = 4.5795e5;
        x_to_beta = pow(x, 2./3.);
    }
    else if (134.43 < d <= 1511.64)
    {
        alpha = 4962.0;
        x_to_beta = pow(x, 1./3.);
    }
    else if (1511.64 < d <= 3477.84)
    {
        alpha = 1732.0;
        x_to_beta = pow(x, 1./6.);
    }
    else
    {
        alpha = 917.0;
        x_to_beta = 1.0;
    }

    return (1e-2 * alpha * x_to_beta); // from cm/s -> m/s

}

Droplet & Droplet::operator= (const Droplet & other) {
    if (this == &other)
        return *this;

    _multi = other._multi;
    _rcubed = other._rcubed;
    _solute = other._solute;
    _density = other._density;

    _volume = other._volume;
    _mass = other._mass;

    return *this;
}

bool operator< (const Droplet & d1, const Droplet & d2) {
    return (d1._multi < d2._multi);
}

// Basic string output
std::ostream &operator<<(std::ostream &os, const Droplet &d) {
    os << "Droplet( "
       << "multi=" << d._multi << ", "
       << "radius=" << d.get_radius()*1e6 << " micron"
       << " )";
    return os;
}

File: droplet.f90
module droplet_class

    use constants
    implicit none
    private

    public :: droplet  ! droplet type
    public :: droplet_ ! droplet constructor bound to init_droplet

    integer(kind=ikind), public :: droplet_count = 0

    type :: droplet
        integer(kind=lkind) :: multi
        real(kind=rkind)    :: rcubed
        real(kind=rkind)    :: solute
        real(kind=rkind)    :: density

        real(kind=rkind)    :: radius, volume, mass
        integer(kind=lkind) :: id
    contains
        procedure, public :: set_droplet
        procedure, public :: terminal_v=>calc_terminal_v
    end type droplet

    interface droplet_
        module procedure init_droplet
    end interface

! ----------------------------------------------------------------------------
contains

    ! Constructor
    type(droplet) function init_droplet(multi, rcubed, solute, density)

        integer(kind=lkind), intent(in) :: multi
        real(kind=rkind), intent(in) :: rcubed
        real(kind=rkind), optional, intent(in) :: solute, density

        init_droplet%multi = multi
        init_droplet%rcubed = rcubed

        if ( present(solute) ) then
            init_droplet%solute = solute
        else
            init_droplet%solute = 0.
        endif

        if ( present(density) ) then
            init_droplet%density = density
        else
            init_droplet%density = RHO_WATER
        endif

        init_droplet%radius = rcubed**(THIRD)
        init_droplet%volume = rcubed * 4. / PI / 3.
        init_droplet%mass   = init_droplet%volume * init_droplet%density

        droplet_count = droplet_count + 1_ikind
        init_droplet%id = droplet_count

    end function init_droplet

    ! Mutator - all attributes
    subroutine set_droplet(self, multi, rcubed, solute)

        class(droplet), intent(inout)  :: self
        integer(kind=lkind), intent(in) :: multi
        real(kind=rkind), intent(in)    :: rcubed, solute

        self%multi = multi
        self%rcubed = rcubed
        self%solute = solute

        self%radius = rcubed**THIRD

    end subroutine set_droplet

    function calc_terminal_v(self) result(terminal_v)
        class(droplet), intent(in) :: self
        real(kind=rkind) :: terminal_v

        real(kind=rkind) :: alpha, r, d, x, x_to_beta
        r = self%radius
        d = 2.*r*1e6 ! diameter, m-> micron
        x = self%mass * 1e3 ! mass, kg -> g

        if ( d <= 134.43 ) then
            alpha = 4.5795e5
            x_to_beta = x**(2*THIRD)
        else if ( (134.43 < d) .and. (d <= 1511.64) ) then
            alpha = 4962.0
            x_to_beta = x**(THIRD)
        else if ( (1511.64 < d) .and. (d <= 3477.84) ) then
            alpha = 1732.0
            x_to_beta = x**(THIRD/2)
        else
            alpha = 917.0
            x_to_beta = 1.0
        end if

        terminal_v = 1e-2 * alpha * x_to_beta ! cm/s -> m/s

    end function calc_terminal_v

end module
